### PR TITLE
feat(timeseries): add subquery support to PromQL evaluator

### LINKED ISSUE #81: support promql subqueries from tsdb
## Summary

Add support for subqueries to the timeseries promql evaluator. This will allow running queries with expressions of the form `<instant_query> '[' <range> ':' [ <resolution> ] ']' [ offset <duration> ]`. Subqueries are explained in detail in the prometheus docs and in the subquery proposal:
- https://prometheus.io/blog/2019/01/28/subquery-support/
- https://prometheus.io/docs/prometheus/latest/querying/basics/#subquery

## Motivation

Needed to support the full PromQL spec.

## Proposed Solution

The evaluator already has a branch for handling subqueries, it's just stubbed with `!todo()`. We need to fill that branch in by executing the subquery the way that `query_range` executes - e.g. by iterating over all the evaluation times and calling back into the evaluator with the specified instant query.

### PR DESCRIPTION
**Pull request by samsond**
Changes:
- Add Timestamp type for clean i64 millisecond arithmetic
- Implement evaluate_subquery with step alignment and time modifiers
- Refactor evaluator to use Timestamp internally (SystemTime only at entry)
- Support @ modifier and offset on subqueries
- Handle negative timestamps for subqueries before UNIX_EPOCH

Tests:
- Add subquery.test with 30+ test cases
- Update at_modifier.test to enable subquery tests
- Ignore 2 edge cases: counter-to-zero extrapolation, native histograms

## Related Issues

Fixes #81 

## Checklist

- [x] Tests added/updated
- [x] `cargo fmt` and `cargo clippy` pass
- [ ] Documentation updated (if applicable)

### CODE REVIEW

**Comment by samsond** [comment]
### apply_time_modifiers:

Originally, apply_time_modifiers was implemented specifically for VectorSelector, since that was the only expression type using @ and offset.

While adding subquery support, it became clear that time modifiers are shared semantics across multiple expression types, not just VectorSelector.

I refactored the function to accept only the minimal required fields (@ and offset) instead of coupling it to a specific expression type. This keeps the logic reusable, reduces coupling, and makes it easier to extend to other expressions that support time modifiers.

### dsl.rs:
Previously, when the parser encountered ignore, it would still attempt to parse commands inside the ignored section (e.g., load) and fail, even though that block should be skipped entirely.

The issue was that command parsing was happening before checking the ignore/resume state.

The fix ensures that the ignore/resume logic is evaluated first. Everything between ignore and resume is now treated as raw text and skipped, regardless of what commands appear inside. This aligns the behavior with the intended DSL semantics.

### Subquery Evaluation (Review Focus)

Most of the change affects subquery evaluation and how time modifiers interact with it.

Nested subqueries were tested and are passing. Time modifier semantics remain unchanged, but please review:

- Evaluation timestamp propagation
- Step alignment (`div_euclid`)
- Interaction between `@`, `offset`, and subquery range

**Comment by samsond** [comment]
Note to self: update the tracking issues for subquery tests after this PR is approved — https://github.com/opendata-oss/opendata/issues/185

File: timeseries/src/promql/promqltest/testdata/subquery.test [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+# Only one value between 10s and 20s, 2@14
+eval instant at 20s min_over_time(metric_total[10s])
+  {} 2
+
+# min(2@20)
+eval instant at 20s min_over_time(metric_total[15s:10s])
+  {} 1
+
+# TODO: Nested subquery with rate() shows ~5x difference due to extrapolation issues.
```
Comment (agavra) [inline timeseries/src/promql/promqltest/testdata/subquery.test]: is this our TODO or a prometheus TODO?

File: timeseries/src/promql/evaluator.rs [hunk 2]
```diff
@@ -596,23 +600,136 @@ impl<'reader, R: QueryReader> Evaluator<'reader, R> {
...
+                labels_key.sort();
+
+                let values = series_map.entry(labels_key).or_default();
+                values.push(Sample {
+                    timestamp_ms: current_time_ms,
+                    value: sample.value,
+                });
+            }
+
+            current_time_ms += step_ms;
```
Comment (agavra) [inline timeseries/src/promql/evaluator.rs]: nit: can this be done as a for loop? a bit easier to read

File: timeseries/src/promql/evaluator.rs [hunk 3]
```diff
@@ -596,23 +600,136 @@ impl<'reader, R: QueryReader> Evaluator<'reader, R> {
...
+
+            for sample in samples {
+                let mut labels_key: Vec<Label> = sample
+                    .labels
+                    .iter()
+                    .map(|(k, v)| Label {
+                        name: k.clone(),
+                        value: v.clone(),
+                    })
+                    .collect();
```
Comment (agavra) [inline timeseries/src/promql/evaluator.rs:705]: this seems like it might be a really hot loop to be doing so much label cloning. probably worth a ticket to figure out how to reduce the cloning overhead here (especially given the labels are probably going to be the same for lots of samples). Maybe maintaining a first map from Unordered Labels -> Ordered Labels so we only create it once (and most of them will probably have the same order of labels).

File: timeseries/src/promql/promqltest/testdata/subquery.test (hunk 1 above)
Comment (samsond) [reply timeseries/src/promql/promqltest/testdata/subquery.test]: Good catch, I'll update the comment to clarify that these gaps in our current implementation

File: timeseries/src/promql/evaluator.rs (hunk 2 above)
Comment (samsond) [reply timeseries/src/promql/evaluator.rs]: cool, I will update this as well.

File: timeseries/src/promql/evaluator.rs (hunk 3 above)
Comment (samsond) [reply timeseries/src/promql/evaluator.rs:705]: Makes sense. Since labels are stable per series, we can maintain a small cache from unordered label sets to their canonical sorted representation and avoid repeated sorting in the subquery step loop. 

I've opened a follow-up ticket to track this optimization: https://github.com/opendata-oss/opendata/issues/242

**Comment by agavra** [comment]
Changes looks good, do you want to hit the button now that you've got write access? 😆 🚀
